titan-compiler/driver.lua

```diff
@@ -0,0 +1,153 @@
+local c_compiler = require "titan-compiler.c_compiler"
+local checker = require "titan-compiler.checker"
+local coder = require "titan-compiler.coder"
+local parser = require "titan-compiler.parser"
+local scope_analysis = require "titan-compiler.scope_analysis"
+local upvalues = require "titan-compiler.upvalues"
+local util = require "titan-compiler.util"
+
+local driver = {}
+
+local function step_index(steps, name)
+    for i, step in ipairs(steps) do
+        if name == step.name then
+            return i
+        end
+    end
+    error("invalid step name "..name)
+end
+
+local ast_passes = {
+    { name = "scope_analysis", f = scope_analysis.bind_names },
+    { name = "checker",        f = checker.check },
+    { name = "upvalues",       f = upvalues.analyze },
+}
+local last_ast_pass = ast_passes[#ast_passes].name
+
+local function compile_pallene_to_ast(pallene_filename, stop_after)
+    local err, errs
+
+    local input
```

Comment:
The main motivator for this was the `ast` variable. For example, in line 42 we have:

    for i = 1, stop_i do
        local pass = ast_passes[i]
        ast, errs = pass.f(ast)
        if not ast then break end
    end

I wanted to reuse the `ast` variable as the loop went on, so that meant that `local ast, errs` wasn't an option. In the end, I felt that declaring things on the top like C was a lesser evil.

Reply:
I think we could do this only when necessary.